### PR TITLE
PDA Messaging

### PR DESCRIPTION
**Pull request by angelofallars**
## About the PR

**Port of Pull Request https://github.com/DeltaV-Station/Delta-v/pull/2362**

Original Message:

> adds pda messaging
> 
> each id card gets assigned a random 4 digit number using nameidentifier system (same system used for things like mobs), this number is used as your chat id rather than anything player facing like character names
> 
> the app ui has a left panel showing all your contacts with unread indicators, and a right panel showing the current chat conversation with bubble messages like a phone app, messages show in grey for others and green for your own messages
> 
> messages are stored directly on the id card component as two dictionaries - one for recipients (storing name/job/unread status) and one for the actual message history, both sender and receiver keep full copies of conversations
> 
> to prevent metagaming the system checks if both pdas have active telecomms connection and are on same station (unless using long range radio channel), failed messages show with a red error indicator
> 
> the detective's logprobe can scan an ID card and "connect" to it, allowing them to spy on people and view their recipients and messages in real time
> 
> interesting mechanics include:
> 
> - agent IDs can be used to impersonate others (they copy message conversation when you scan an ID, you can also change your number)
> - microwaving an id has 10% chance to wipe all messages, or randomly scramble message text and reassign conversations between contacts
> - message delivery requires working telecomms and respects radio blocking
> - notifications show truncated message previews up to 64 chars
> - contacts remember job titles from id cards (only if you message someone you dont know first)
> - messages store timestamps for rebuilding conversation order
> - supports up to 50 contacts per card
> - messages limited to 256 chars

## Why / Balance

Original Message:

> it's cool

## Media

Original Message:

<details><summary>Click me!</summary>

hiiiii

image

image

image
image
image
image

(microwaved)

image

</details>

## Requirements

- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an ingame showcase.

## Breaking changes

**Changelog**

:cl: MilonPL
- add: Added PDA messaging!

### CODE REVIEW

**Review by whateverusername0 — CHANGES_REQUESTED** [review]
we've been marking fork folders with an underscore for better navigation please do that here as well
image

**Comment by deltanedas** [comment]
thats how it is on deltav tbh i might change it to _ and break everything in the future

**Comment by angelofallars** [comment]
> we've been marking fork folders with an underscore for better navigation please do that here as well image

I moved most of the files to the `_DeltaV` namespace, but some changes in `Resources/` were modifications to existing `DeltaV` files that other files unrelated to PDA messaging also rely on. Should I leave them as is or do you want me to also move them entirely to `_DeltaV`?

**Comment by Piras314** [comment]
We can fix the existing DeltaV folder (non _) stuff in a fix PR later.

File: Content.Client/_DeltaV/CartridgeLoader/Cartridges/NanoChatUiFragment.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+
+            // Show character count when over limit
+            CharacterCount.Visible = length > MaxMessageLength;
+            if (length > MaxMessageLength)
+            {
+                CharacterCount.Text = Loc.GetString("nano-chat-message-too-long",
+                    ("current", length),
+                    ("max", MaxMessageLength));
+                CharacterCount.StyleClasses.Add("LabelDanger");
+            }
```
Comment (Aviu00) [inline Content.Client/_DeltaV/CartridgeLoader/Cartridges/NanoChatUiFragment.xaml.cs:75]: I don't like how it is checked clientside only. Message should be truncated to `MaxMessageLenght` on server side as well.

File: Content.Client/_DeltaV/CartridgeLoader/Cartridges/NanoChatUiFragment.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        };
+
```
Comment (AgentePanela) [inline Content.Client/_DeltaV/CartridgeLoader/Cartridges/NanoChatUiFragment.xaml.cs:77]:
```suggestion

MessageInput.OnTextEntered += _ => SendMessage(); // Send message when press enter
```

i add this to my downstream when a i port the messaging

Comment (angelofallars) [reply Content.Client/_DeltaV/CartridgeLoader/Cartridges/NanoChatUiFragment.xaml.cs:75]: Okay I'll add this too

**Comment by angelofallars** [comment]
Server-side validation implemented with [c063001](https://github.com/Goob-Station/Goob-Station/pull/1269/commits/c063001ca97df3cd4ea69639b8f9c0cb9e51fc90). Message content when sending a message, and the recipient's name and job title when creating a new chat will now be trimmed of leading/trailing whitespaces and truncated when they go past their respective maximum length.

File: Content.Client/_DeltaV/CartridgeLoader/Cartridges/NanoChatUiFragment.xaml.cs (hunk 2 above)
Comment (angelofallars) [reply Content.Client/_DeltaV/CartridgeLoader/Cartridges/NanoChatUiFragment.xaml.cs:75]: resolved with [0aec325](https://github.com/Goob-Station/Goob-Station/pull/1269/commits/0aec3258bb1dded4a680f82373932a6c4bc820d3)

File: Content.Client/_DeltaV/CartridgeLoader/Cartridges/NanoChatUiFragment.xaml.cs (hunk 1 above)
Comment (angelofallars) [reply Content.Client/_DeltaV/CartridgeLoader/Cartridges/NanoChatUiFragment.xaml.cs]: resolved with [c063001](https://github.com/Goob-Station/Goob-Station/pull/1269/commits/c063001ca97df3cd4ea69639b8f9c0cb9e51fc90)

**Review by Piras314 — APPROVED** [review]
LGTM

**Pull request merged**